### PR TITLE
Test App Updates

### PR DESCRIPTION
**Pull request by simpsont-oci**
Improves "random" sample generation in order to test more of dynamic data. Perhaps unsurprisingly, causes SEGV in the monitor for some generated samples when subscribing to the unmanaged test app (using dynamic data).

Given the test app IDL, the following samples have all been observed to cause SEGV:

```
 - {"origin":"unmanaged_testapp-lumoria-100374","count":1,"ct":{"bt":{"u8":1,"i8":1,"us":1,"s":1,"ul":1,"l":1,"ull":1,"ll":1,"f":1.0,"d":1.0,"b":true,"o":1,"c":"b","wc":"Y","str":""},"cuts":[{"$discriminator":"one","str":"A complex union string: 1216115135"},{"$discriminator":"four","tns":[]}]}}
 - {"origin":"unmanaged_testapp-lumoria-100426","count":2,"ct":{"bt":{"u8":2,"i8":2,"us":2,"s":2,"ul":2,"l":2,"ull":2,"ll":2,"f":2.0,"d":2.0,"b":false,"o":2,"c":"c","wc":"X","str":""},"cuts":[{"$discriminator":"one","str":"A complex union string: 1328248073"},{"$discriminator":"four","tns":[]}]}}
 - {"origin":"unmanaged_testapp-lumoria-100584","count":0,"ct":{"bt":{"u8":0,"i8":0,"us":0,"s":0,"ul":0,"l":0,"ull":0,"ll":0,"f":0.0,"d":0.0,"b":false,"o":0,"c":"a","wc":"Z","str":""},"cuts":[{"$discriminator":"two","ut":{"$discriminator":"one","str":"A union string: 1978156891"}},{"$discriminator":"three","tn":{"et":"three","ut":{"$discriminator":"four","btss":[[]]},"tns":[{"et":"four","ut":{"$discriminator":"two","bt":{"u8":64,"i8":87,"us":32458,"s":15036,"ul":1349045358,"l":406939873,"ull":1961668412,"ll":2068534139,"f":3123291904.0,"d":2162427736.0,"b":false,"o":42,"c":"e","wc":"M","str":"A basic string 2498535724"}},"tns":[]}]}}]}}
 - {"origin":"unmanaged_testapp-lumoria-100663","count":0,"ct":{"bt":{"u8":0,"i8":0,"us":0,"s":0,"ul":0,"l":0,"ull":0,"ll":0,"f":0.0,"d":0.0,"b":false,"o":0,"c":"a","wc":"Z","str":""},"cuts":[{"$discriminator":"four","tns":[]},{"$discriminator":"four","tns":[{"et":"four","ut":{"$discriminator":"two","bt":{"u8":156,"i8":118,"us":28891,"s":6011,"ul":4147023377,"l":1656741056,"ull":1456839787,"ll":2746201045,"f":414862368.0,"d":941259102.0,"b":true,"o":87,"c":"u","wc":"M","str":"A basic string 1861234030"}},"tns":[]}]}]}}
 - {"origin":"unmanaged_testapp-lumoria-100711","count":0,"ct":{"bt":{"u8":0,"i8":0,"us":0,"s":0,"ul":0,"l":0,"ull":0,"ll":0,"f":0.0,"d":0.0,"b":false,"o":0,"c":"a","wc":"Z","str":""},"cuts":[{"$discriminator":"three","tn":{"et":"three","ut":{"$discriminator":"three","bts":[{"u8":233,"i8":87,"us":38893,"s":19840,"ul":278126418,"l":1791332590,"ull":3253152003,"ll":1099833389,"f":3664764928.0,"d":3401278335.0,"b":false,"o":75,"c":"n","wc":"T","str":"A basic string 3643114909"}]},"tns":[]}},{"$discriminator":"one","str":"A complex union string: 1347159026"}]}}
 - {"origin":"unmanaged_testapp-lumoria-101031","count":2,"ct":{"bt":{"u8":2,"i8":2,"us":2,"s":2,"ul":2,"l":2,"ull":2,"ll":2,"f":2.0,"d":2.0,"b":false,"o":2,"c":"c","wc":"X","str":""},"cuts":[{"$discriminator":"one","str":"A complex union string: 1543224235"},{"$discriminator":"three","tn":{"et":"two","ut":{"$discriminator":"one","str":"A union string: 3453233805"},"tns":[]}}]}}
```

### CODE REVIEW

File: src/topic_table_model.cpp [hunk 1]
```diff
@@ -793,8 +793,10 @@ void TopicTableModel::parseAggregated(const DDS::DynamicData_var& data, const st
             continue;
         }
 
-	std::string scoped_member_name = namePrefix.empty() ? md->name() : namePrefix + "." + md->name();
-        const OpenDDS::XTypes::TypeKind member_tk = OpenDDS::XTypes::get_base_type(md->type())->get_kind();
+        std::string scoped_member_name = namePrefix.empty() ? md->name() : namePrefix + "." + md->name();
+        const DDS::DynamicType_var type = md->type();
+        const DDS::DynamicType_var base_type = OpenDDS::XTypes::get_base_type(type.in());
```
Comment (mitza-oci) [inline src/topic_table_model.cpp:798]:
```suggestion
        const DDS::DynamicType_var base_type = OpenDDS::XTypes::get_base_type(md->type());
```
This is one of the odd things about IDL Valuetypes.  The "accessors" like `md->type()` do not increase the refcount on the object being returned.  (IDL Unions are the same.)

File: test/test.idl [hunk 2]
```diff
@@ -3,8 +3,8 @@
 module test {
 
 struct BasicTypes {
-  uint8 u8;
-  int8 i8;
+  unsigned short u8;
```
Comment (mitza-oci) [inline test/test.idl:6]: change name to u16 or change type

**Review by mitza-oci — CHANGES_REQUESTED** [review]

File: test/json_conversion.h [hunk 3]
```diff
@@ -0,0 +1,33 @@
+#pragma once
+
+#include <iostream>
+
+#include <dds/DCPS/RapidJsonWrapper.h>
```
Comment (sonndinh) [inline test/json_conversion.h:5]: Should the order of these includes be switched?

File: test/json_conversion.h [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+//bool json_2_idl(std::istream& is, IDL_Type& idl_value) {
+//  rapidjson::IStreamWrapper isw(is);
+//  return OpenDDS::DCPS::from_json(idl_value, isw);
+//}
+
+template<typename IDL_Type, typename Writer_Type = rapidjson::Writer<rapidjson::OStreamWrapper> >
+bool idl_2_json(const IDL_Type& idl_value, std::ostream& os,
+    int max_decimal_places = Writer_Type::kDefaultMaxDecimalPlaces) {
+  rapidjson::OStreamWrapper osw(os);
+  Writer_Type writer(osw);
```
Comment (sonndinh) [inline test/json_conversion.h:24]:
```suggestion
bool idl_2_json(const IDL_Type& idl_value, std::ostream& os,
    int max_decimal_places = Writer_Type::kDefaultMaxDecimalPlaces)
{
  rapidjson::OStreamWrapper osw(os);
  Writer_Type writer(osw);
```

**Review by mitza-oci — APPROVED** [review]

**Review by sonndinh — APPROVED** [review]

**Pull request merged**